### PR TITLE
add IMAP auth support

### PR DESCRIPTION
**Pull request by akoscomp**
I add code for IMAP auth support via php imap_open function.

### CODE REVIEW

File: mailscanner/functions.php [hunk 1]
```diff
@@ -2972,6 +2972,46 @@ function ldap_authenticate($username, $password)
...
+ * @param string $username
+ * @param string $password
+ * @return null|string
+ */
+function imap_authenticate($username, $password)
+{
+    $username = strtolower($username);
+    $email = $username;
+
+    if (!filter_var($email, FILTER_VALIDATE_EMAIL)) {
```
Comment (endelwar) [inline mailscanner/functions.php]: You can use `$username` directly without creating `$email` var

File: mailscanner/functions.php [hunk 2]
```diff
@@ -2972,6 +2972,46 @@ function ldap_authenticate($username, $password)
...
+        $mbox = imap_open(IMAP_HOST, $username, $password, NULL, 0);
+
+        if ( false === $mbox ) {
+            //auth faild
+            return null;
+        }
+
+        $sql = sprintf('SELECT username FROM users WHERE username = %s', quote_smart($email));
+        $sth = dbquery($sql);
+        if ($sth->num_rows === 0) {
```
Comment (endelwar) [inline mailscanner/functions.php]: Add a `IMAP_AUTOCREATE_VALID_USER` boolean constant to enable this behaviour, set it false by default

File: mailscanner/checklogin.php [hunk 3]
```diff
@@ -68,6 +68,24 @@
     }
 }
 
+if (
```
Comment (endelwar) [inline mailscanner/checklogin.php]: this `if` should be chained with the one on line 53, using an `elseif`, and `else` on line 60 should also set `$_SESSION['user_imap'] = false;`

**Comment by Skywalker-11** [comment]
You should also add a check if user_imap is used in the user_manager.php:
https://github.com/mailwatch/MailWatch/blob/0140fefb15ea90d68803d52ce124c4b9339f3607/mailscanner/user_manager.php#L781

**Comment by Skywalker-11** [comment]
Maybe we can also make the parameters optional so users don't necessarily need to update their config by changing
```
} elseif (
     (USE_IMAP === true) &&
     (($result = imap_authenticate($myusername, $mypassword)) !== null)
 ) {
``` 
to 
```
} elseif (
     defined('USE_IMAP') &&
     (USE_IMAP === true) &&
     (($result = imap_authenticate($myusername, $mypassword)) !== null)
 ) {
```
And moving `USE_IMAP` and `IMAP_HOST` from the $needed array to the optional array eg:
```
...
  'ENABLE_SUPER_DOMAIN_ADMINS' => array('description' => 'allows domain admins to change domain admins from the same domain'),
  'USE_IMAP' => array('description' => 'use IMAP for user authentication'),
  'IMAP_HOST' => array('description' => 'IMAP host to be used for user authentication')
);
```

**Comment by endelwar** [comment]
I've added @Skywalker-11 suggestions to the code, this need to be tested, specially the optional autocreate user from imap login

**Pull request merged**